patroni/postgresql/validator.py

```diff
@@ -128,456 +142,369 @@ def transform(self, name: str, value: Optional[Any]) -> Optional[Any]:
+        GUC. Used for a couple purposes:
+        * Disallow writing GUCs to ``postgresql.conf`` (or ``recovery.conf``) that does not exist in Postgres *version*;
+        * Avoid ignoring GUC *name* if it does not have a validator in *validators*, but is a valid GUC in Postgres
+            *version*.
+
+    :returns: the return value may be one among:
+        * *value* transformed to the expected format for GUC *name* in Postgres *version*, if *name* is present in
+            *available_gucs* and has a validator in *validators* for the corresponding Postgres *version*; or
+        * The own *value* if *name* is present in *available_gucs* but not in *validators*; or
+        * ``None`` if *name* is not present in *available_gucs*.
+    """
+    if name in available_gucs:
+        name_validators: Tuple[_Transformable, ...] = validators.get(name, ())
+        if name_validators:
+            for validator in name_validators:
+                if version >= validator.version_from and\
+                        (validator.version_till is None or version < validator.version_till):
+                    return validator.transform(name, value)
+        # Ideally we should have a validator in *validators*. However, if none is available, we will not discard a
+        # setting that exists in Postgres *version*, but rather allow the value with no validation.
+        return value
     logger.warning('Removing unexpected parameter=%s value=%s from the config', name, value)
 
 
-def transform_postgresql_parameter_value(version: int, name: str, value: Any) -> Optional[Any]:
+def transform_postgresql_parameter_value(version: int, name: str, value: Any,
+                                         available_gucs: CaseInsensitiveSet) -> Optional[Any]:
+    """Validate *value* of GUC *name* for Postgres *version* using ``parameters`` and *available_gucs*.
+
+    :param version: Postgres version to validate the GUC against.
+    :param name: name of the Postgres GUC.
+    :param value: value of the Postgres GUC.
+    :param available_gucs: a set of all GUCs available in Postgres *version*. Each item is the name of a Postgres
+        GUC. Used for a couple purposes:
+        * Disallow writing GUCs to ``postgresql.conf`` that does not exist in Postgres *version*;
+        * Avoid ignoring GUC *name* if it does not have a validator in ``parameters``, but is a valid GUC in Postgres
+            *version*.
+
+    :returns: The return value may be one among
+        * The original *value* if *name* seems to be an extension GUC (contains a period '.'); or
+        * ``None`` if **name** is a recovery GUC; or
+        * *value* transformed to the expected format for GUC *name* in Postgres *version* using validators defined in
+            ``parameters``. Can also return ``None``. See :func:`_transform_parameter_value`.
+    """
     if '.' in name:
-        return value
+        if name not in parameters:
```

Comment:
`if '.' in name and name not in parameters:`

Reply:
Applied, thanks :)